### DISCUSSION TITLE
admin sanitation of POST GET

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#5836] [Ideas]
While fettling with category-product listing, **some** POST/GET are pre-cast to integers here
https://github.com/zencart/zencart/blob/d46e790dbab89ac4aab94496784632a65c17fe8d/admin/includes/init_includes/init_sanitize.php#L171-L188

Other POST/GET in this file are not pre-sanitized , the usual in-file cleaning or casting is done, eg:
```
if (isset($_POST['categories_id']) && ($_POST['categories_id'] !== $_POST['move_to_category_id'])) {
        $categories_id = zen_db_prepare_input($_POST['categories_id']);
        $new_parent_id = zen_db_prepare_input($_POST['move_to_category_id']);
```
where $_POST['categories_id' IS pre-cast to integer but $_POST['move_to_category_id'] is not pre-sanitized at all.

Is the philosophy to move sanitizing out of a file and into the init?

So $_POST['move_to_category_id'] should also go in the $goup array and 
`$new_parent_id = zen_db_prepare_input($_POST['move_to_category_id'])`;
becomes
`$new_parent_id = $_POST['move_to_category_id');`

**Comment by mc12345678** [comment]
While not an answer to what seems to be your real question, in a default store, `$_POST['move_to_category_id']` *DOES* get (pre-)sanitized, but not specifically to an integer, which may become an issue within PHP operations. 

Because it is not specifically identified in the init_sanitize code it is strict sanitized. In your testing did you find that it bypassed this particular POST key? 

That any/all presanitization eliminates the need to use `zen_db_prepare_input` on the admin side for `$_POST` and `$_GET` variables then becomes the focus for discussion except where PHP has issue with operations on those variables with them not cast as integers or the desired type.

**Comment by torvista** [comment]
> Because it is not specifically identified in the init_sanitize code it is strict sanitized.

Where?

**Comment by mc12345678** [comment]
> > Because it is not specifically identified in the init_sanitize code it is strict sanitized.
> 
> Where?

At start of `admin/includes/init_includes/init_sanitize.php` there is an assignment to specifically perform strict sanitization unless the constant is defined at which point strict sanitization is as previously defined. 

There is then the call near the end of the file to run sanitizers, this invokes filtering of both keys and the related values in a strict manner. 

Such information would be logged if `DO_DEBUG_SANITIZATION` was strictly set to `true`. While such logging wouldn't identify the line number in the code that performed the sanitization, it is expected to show that both the key and the value was strictly sanitized. Again I note that such sanitization would not force the value to be an integer which may cause an issue in php code depending on the involved operation. The value would be sanitized using `htmlspecialchars` enforcing double encoding with the flag of `ENT_COMPAT` which converts double-quotes, but leaves single-quotes alone.

**Comment by torvista** [comment]
Thanks for the info, although the correct answer to "where?" was RTM:
https://docs.zen-cart.com/dev/code/admin_sanitization/

The original question remains

> Is the philosophy to move sanitizing out of a file and into the init?

**Comment by zcwilt** [comment]
And at some point we need to have a conversation as to why we do this and how we stop having to do this.

**Comment by lat9** [comment]
@zcwilt, what is `this`?

**Comment by zcwilt** [comment]
security through obscurity.

This include admin sanitization and obscured admin directories.

**Comment by zcwilt** [comment]
@lat9 will put together some notes around this

**Comment by proseLA** [comment]
> And at some point we need to have a conversation as to why we do this and how we stop having to do this.


> security through obscurity.
> 
> This include admin sanitization and obscured admin directories.

i could not agree more.

**Comment by lat9** [comment]
As noted in the issue mentioned by @torvista, there's also been some [discussion](https://github.com/zencart/zencart/discussions/4824) of this handling over the past year.
